enhancements/baremetal/ztp-metal3.md

```diff
@@ -0,0 +1,249 @@
+
+### Risks and Mitigations
+
+- More containers will be downloaded on the ramdisk start-up. We'll need to
+  account for more RAM and a longer start-up time.
+- A user with editing access to `InfraEnv` resources can potentially break
+  the Ironic agent configuration. Short-term, we'll probably document this
+  as a known issue. Long-term, we may need to update `InfraEnv` with a new
+  field.
+
+## Design Details
+
+### Open Questions
+
+- ICC will store the user's pull secret directly to the Ignition override
+  field the `InfraEnv` resource (as a base64 encoded text, not as a reference
+  to secret). Is it acceptable?
+- Is there a chance of version mismatch between the CoreOS `initrd` from the
+  assisted service and the root filesystem provided by
+  [coreos-downloader][coreos-downloader]?
+
+### Test Plan
+
+The feature will be covered by the existing ZTP e2e tests.
+
+### Graduation Criteria
+
+This work will be GA immediately since there is no phasing possible or planned.
+Only ZTP will be affected.
+
+#### Dev Preview -> Tech Preview
+
+N/A
+
+#### Tech Preview -> GA
+
+N/A
+
+#### Removing a deprecated feature
+
+N/A
+
+### Upgrade / Downgrade Strategy
+
+This change does not introduce new resources, so upgrading or downgrading them
+is not a concern.
+
+For the new flow to apply, it must be explicitly requested through the
+`BareMetalHost` API. Otherwise, the old flow is used.
+
+There are no plans to remove the code that supports the *live ISO* flow from
+Ironic, Metal3 or OpenShift.
+
+### Version Skew Strategy
+
+The image-customization-controller will need to be updated before the new flow
+will be usable. Given that ZTP releases trail OpenShift releases (and are tied
+to them), it should not become an issue.
+
+### Operational Aspects of API Extensions
+
+#### Failure Modes
+
+- When image building fails in an `InfraEnv` resource, this failure will be
+  propagated to the baremetal-operator's `PreprovisioningImage` resource via
+  conditions. The whole provisioning process will be paused until the failure
+  is fixed.
+
+#### Support Procedures
+
+- The `image-customization-controller` logs will be the central place to
+  triage issues related to image building. Such issues will be highlighted
+  by a failure condition on the `PreprovisioningImage` resource.
+- The Ironic ramdisk logs (published via a special container
+  `ironic-ramdisk-logs`) will be useful to debug the coordination between
+  the two agents.
+
+## Implementation History
+
+- [Updated image building
+  process](https://github.com/openshift/image-customization-controller/pull/42)
+- [Ironic agent
+  plugin](https://github.com/openshift/ironic-agent-image/pull/40)
+
+## Drawbacks
+
+- The converged flow is more complex than each of the old flows.
+- The image-customization-controller becomes aware of (although does not
+  require) the assisted-service API.
+
+## Alternatives
+
+- Keep both flows separate, implement all required feature twice.
+
```

Comment:
An alternative we discussed is for the ICC to retrieve the Ignition from assisted-service and build the image including the IPA config itself, instead of the passing the IPA config to assisted-service and having that assisted-image-service build the image.

(A drawback to this is that minimal ISO support has not yet been added to ICC.)